### ISSUE TITLE
Remote write: treat http status 408 as retriable

### ISSUE THREAD

**Issue by krajorama** [prometheus/prometheus#12409]
### Proposal

The remote metrics database may return [408 Request Timeout](https://http.dev/408). Currently this is not retried as it should be.

As a workaround the server can instead return [504 Gateway Timeout](https://http.dev/504) to force retry, but this is misleading in logs.

**Comment by bboreham** [comment]
Since I happened to be looking for references, this seems more authoritative: https://httpwg.org/specs/rfc9110.html#status.408
> The 408 (Request Timeout) status code indicates that the server did not receive a complete request message within the time that it was prepared to wait.

> If the client has an outstanding request in transit, it MAY repeat that request. If the current connection is not usable (e.g., as it would be in HTTP/1.1 because request delimitation is lost), a new connection will be used.

(I agree it would be good for Prometheus to retry)

**Comment by roidelapluie** [comment]
Should we generalize our treatement of 429 and provide a configurable array of error codes to retry in the 4xx range?

**Comment by bboreham** [comment]
Personally I think neither should be configurable; Prometheus should retry on both. 

Originally retry-on-429 was optional because data would queue up in memory and crash the process if the overload persisted. But since we moved to use the WAL this is not a danger and the oldest data is deleted after WAL checkpoint (by default every 2 hours). 

It’s a change, but to fix a bug.

**Comment by roidelapluie** [comment]
No, originally we went for a config option because it is a change of
policy. We have documented extensively on every media that Prometheus 2.x
does not retry 4xx errors.

Le sam. 17 juin 2023, 13:12, Bryan Boreham ***@***.***> a
écrit :

**Comment by 56quarters** [comment]
> Should we generalize our treatement of 429 and provide a configurable array of error codes to retry in the 4xx range?

From a quick look at the remote write code, it doesn't seem like very many changes to add another `retry_on_429`-style flag and thread that through everywhere that's needed.

**Comment by beorn7** [comment]
Hello from the bug scrub!

@krajorama @bboreham @bwplotka @alexgreenbank what's the current state of this? [PRW2.0](https://prometheus.io/docs/specs/remote_write_spec_2_0/) talks a lot about the retries. Is it safe to assume that we can close this?

**Comment by bwplotka** [comment]
Retrying on 429 is part of the 2.0 spec. 408 is not and no plans for that. 

Is this still useful to consider @krajorama ?

We could still consider such a change perhaps or do it only in Prometheus or only in 2.1.

**Comment by krajorama** [comment]
I'm not in the loop any more about this. Let me ask someone.
